### PR TITLE
Schedule styling tweaks [WIP]

### PR DESCRIPTION
**Pull request by kamiyo**
code review/suggestions/debugging please

### CODE REVIEW

**Comment by kamiyo** [comment]
To see new stuff, go to schedule, click on one event, should route to single event details.

Issues:
1. If you click Schedule again while already on Schedule, the page disappears.
2. If you load a single event page through the address bar (i.e. localhost:8000/schedule/2017-08-21) and then click Schedule, the list doesn't appear (even though it exists in the DOM, somehow the positioning is messed up). I don't think the clone element in the transition animation container is correct...
3. If you go to /schedule, then click an item, then go back or click Schedule on the nav, then try to go to a single item again, the whole site reloads... maybe an issue with how the history is being managed...
4. Not sure if the scroll-saving is working correctly (when you leave the list to go to a single item, when you go back to the list, it should save scroll location.

Not really related:
1. react-dates calendar is kinda limited, for example you can't force change the current month (have to do that thru UX), so might have to change calendars to something more manipulateable.

...maybe more

**Comment by ayc92** [comment]
You should add `.vscode` to `.gitignore`

File: web/js/components/App/NavBar/NavBarLink.jsx [hunk 1]
```diff
@@ -2,9 +2,15 @@ import '@/less/App/NavBar/sub-nav.less';
 
 import React from 'react';
 import ReactDOM from 'react-dom';
-import {Link, IndexLink, withRouter} from 'react-router';
+import { Link, IndexLink, withRouter} from 'react-router';
```
Comment (ayc92) [inline web/js/components/App/NavBar/NavBarLink.jsx]: Missing space before closing '}'

File: web/js/components/App/NavBar/NavBarLink.jsx [hunk 2]
```diff
@@ -50,10 +56,9 @@ class NavBarLink extends React.Component {
                 active = 'active';
             }
         } else {
-            if (this.props.router.isActive(link, false) && !this.props.showSub) {
+            if (this.props.link == this.props.router.location.pathname.split('/')[1] && !this.props.showSub) {
```
Comment (ayc92) [inline web/js/components/App/NavBar/NavBarLink.jsx]: Use `===` for strings.

Comment (ayc92) [inline web/js/components/App/NavBar/NavBarLink.jsx]: Also, what exactly is this logic doing?

File: web/js/components/App/NavBar/NavBarLink.jsx [hunk 3]
```diff
@@ -62,20 +67,21 @@ class NavBarLink extends React.Component {
...
                             : (
-                                <Link to={'/' + link} onClick={() => this.props.toggleSub(false)} className={active}>
-                                    <div className={highlightClass}></div>
-                                    <div className="hyperlink">{link}</div>
+                                <Link
+                                    to={'/' + link}
+                                    onClick={() => this.props.toggleSub(false)}
+                                    className={active}
+                                >
+                                    <Highlight highlightClass={highlightClass} link={link} />
```
Comment (ayc92) [inline web/js/components/App/NavBar/NavBarLink.jsx]: I think this part can be cleaned up. I feel like `NavBarLink` should just be a dumb component. It should just know whether it should be highlighted (via a `isHighlighted` prop or something). It should be the responsibility of the containing component to pass down the correct props, meaning the parent component should be the one that has the logic of determining which nav link should be highlighted.

File: web/js/components/Schedule/EventItem.jsx [hunk 4]
```diff
@@ -1,41 +1,82 @@
...
+
+const mapDispatchToProps = dispatch => ({
+    storeScroll: scrollTop => {
+        dispatch({type: 'SCHEDULE--STORE_SCROLL', scrollTop});
+    },
+});
+
+export default withRouter(
+    connect(
+        null,
```
Comment (ayc92) [inline web/js/components/Schedule/EventItem.jsx]: What's the use of passing null here? I feel like passing a no-op `() => {}` is better practice, because we expect this to be a function (i.e. it'll probably be invoked at some point).

File: web/js/components/Schedule/EventList.jsx [hunk 5]
```diff
@@ -18,22 +16,32 @@ const cache = new CellMeasurerCache({
...
     }
 
-    componentWillMount() {
-        this.props.fetchEvents();
+    shouldComponentUpdate(nextProps, nextState) {
+        // want to prevent 'scrollToRow' when transitioning between eventList and eventSingle
+        if (nextProps.currentScrollIndex == this.props.currentScrollIndex)
+            if (nextProps.eventItems == this.props.eventItems)
+                return false;
+        return true;
```
Comment (ayc92) [inline web/js/components/Schedule/EventList.jsx]: You can just do:
```
return (
    nextProps.currentScrollIndex !== this.props.currentScrollIndex ||
    nextProps.eventItems != this.props.eventItems
);
```

File: web/js/components/Schedule/EventList.jsx [hunk 6]
```diff
@@ -18,22 +16,32 @@ const cache = new CellMeasurerCache({
...
     }
 
-    componentDidUpdate() { this.List.scrollToRow(this.props.currentScrollIndex || 0); }
+    componentDidUpdate() {
+        this.List.scrollToRow(this.props.currentScrollIndex || 0);
+    }
+
+    componentDidMount() {
+        //without timeout, scrollToPosition does not work correctly
+        setTimeout(() => {this.List.scrollToPosition(this.props.scrollTop);}, 100);
```
Comment (ayc92) [inline web/js/components/Schedule/EventList.jsx]: I think you can just do `setTimeout(() => this.List.scrollToPosition(this.props.scrollTop), 100);`

File: web/js/components/Schedule/EventList.jsx [hunk 7]
```diff
@@ -53,52 +61,33 @@ class ConnectedEventList extends React.Component {
...
-        ));
-        dispatch({ type: 'SCHEDULE--FETCHING_EVENTS' });
-    }
+    scrollTop: state.schedule_eventItems.scrollTop,
 });
 
 export default connect(
     mapStateToProps,
-    mapDispatchToProps
+    null
```
Comment (ayc92) [inline web/js/components/Schedule/EventList.jsx]: Same here as before regarding passing a no-op instead of null.

File: web/js/components/Schedule/EventMonthItem.jsx [hunk 8]
```diff
@@ -2,14 +2,12 @@ import '@/less/Schedule/event-month-item.less';
...
-                </div>
-            </div>
-        );
-    }
-}
+const EventMonthItem = ({ style, month }) => (
+    <div className="event-month-item" style={style}>
+        <div>
+            {month}
+        </div>
```
Comment (ayc92) [inline web/js/components/Schedule/EventMonthItem.jsx]: Is the inner div necessary here? Doesn't seem like it does anything.

File: web/js/components/Schedule/EventSingle.jsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import '@/less/Schedule/event-month-item.less';
+import '@/less/Schedule/event-item.less';
+
+import React from 'react';
+import moment from 'moment-timezone';
+import { connect } from 'react-redux';
+
+import EventMonthItem from '@/js/components/Schedule/EventMonthItem.jsx';
+
+const EventNoExist = ({ date }) => (
```
Comment (ayc92) [inline web/js/components/Schedule/EventSingle.jsx]: Maybe `NoEventsScheduled` or something grammatically correct? :)

File: web/js/components/Schedule/EventSingle.jsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+const mapStateToProps = state => ({
+    eventItems: state.schedule_eventItems.items,
+    date: state.schedule_date.date,
+});
+
+export default connect(
+    mapStateToProps,
+    null
```
Comment (ayc92) [inline web/js/components/Schedule/EventSingle.jsx]: Same here re: passing null

File: web/js/components/Schedule/EventItem.jsx [hunk 11]
```diff
@@ -1,41 +1,82 @@
...
+    <div className="event-item__date-container">
+        <div className="event-item__date">
+            {day}
+        </div>
+    </div>
+);
+
+const EventName = ({ date, name, storeScroll, gridState }) => (
+    <Link
+        to={'/schedule/' + date.format('YYYY-MM-DD')}
```
Comment (ayc92) [inline web/js/components/Schedule/EventItem.jsx]: You can use template strings :)

```
`/schedule/${date.format('YYYY-MM-DD')}`
```

File: web/js/components/Schedule/EventItem.jsx (hunk 4 above)
Comment (kamiyo) [inline web/js/components/Schedule/EventItem.jsx]: this gives an error, or i'm not doing it right. I did:

```
withRouter(
   connect(
      () => {},
      mapDispatchToProps
)(EventItem));
```

And it gives: mapStateToProps() in Connect(EventItem) must return a plain object. Instead received undefined.

Comment (kamiyo) [inline web/js/components/Schedule/EventItem.jsx]: actually per docs: https://github.com/reactjs/react-redux/blob/master/docs/api.md#arguments

should pass null or undefined for mapStateToProps if we don't need it, and if we don't need mapDispatchToProps, we can leave it off, because it injects a default mapDispatchToProps which injects `dispatch` into `this.props`

Comment (kamiyo) [inline web/js/components/Schedule/EventItem.jsx]: oh actually if i do () => ({}) it doesn't give the errors anymore.

File: web/js/components/App/NavBar/NavBarLink.jsx (hunk 3 above)
Comment (kamiyo) [inline web/js/components/App/NavBar/NavBarLink.jsx]: I think we should work on this in another PR.

File: web/js/components/Schedule/EventItem.jsx (hunk 4 above)
Comment (ayc92) [inline web/js/components/Schedule/EventItem.jsx]: Oh, in that case, null is fine then.

File: web/js/components/App/App.jsx [hunk 12]
```diff
@@ -37,23 +37,24 @@ export default class App extends React.Component {
...
-                    enter={{duration: 500, animation: {opacity: 1, translateZ: 0}}}
-                    leave={{duration: 500, animation: {opacity: 0, translateZ: 0}}}
+                    enter={{ duration: 500, animation: { opacity: 1, translateZ: 0 } }}
+                    leave={{ duration: 500, animation: { opacity: 0, translateZ: 0 } }}
                     runOnMount={true}
                 >
                     {React.cloneElement(this.props.children, {
-                        key: this.props.location.pathname
+                        // we don't want animation here to trigger when going from /schedule to /schedule/[date]
+                        key: this.props.location.pathname.split('/')[1]
```
Comment (ayc92) [inline web/js/components/App/App.jsx]: Can you explain what's going on here?

File: web/js/components/App/NavBar/NavBarLink.jsx (hunk 3 above)
Comment (ayc92) [inline web/js/components/App/NavBar/NavBarLink.jsx]: Sounds good to me

File: web/js/components/App/NavBar/NavBarLink.jsx (hunk 2 above)
Comment (ayc92) [inline web/js/components/App/NavBar/NavBarLink.jsx]: Yeah maybe the `NavBarLink` component should take a prop that's like, `baseRoute` or something, which in this case would be "schedule". What do u think?

File: web/js/components/Schedule/EventList.jsx [hunk 13]
```diff
@@ -18,22 +16,32 @@ const cache = new CellMeasurerCache({
...
         super(props, context);
-
         this._renderEventItem = this._renderEventItem.bind(this);
     }
 
-    componentWillMount() {
-        this.props.fetchEvents();
+    shouldComponentUpdate(nextProps, nextState) {
+        // want to prevent 'scrollToRow' when transitioning between eventList and eventSingle
+        if (nextProps.currentScrollIndex == this.props.currentScrollIndex)
```
Comment (ayc92) [inline web/js/components/Schedule/EventList.jsx]: I don't really understand why the list is being re-rendered when you transition from event list -> event single. It should just be unmounted, right?

File: web/js/main.jsx [hunk 14]
```diff
@@ -18,24 +19,33 @@ import Videos from '@/js/components/Media/Videos/Videos.jsx';
...
-                    <Route path='photos' component={Photos} />
+        <Provider store={store}>
+            <Router history={browserHistory}>
+                <Route path='/' component={App}>
+                    <IndexRoute component={Home} />
+                    <Route path='about' component={About} />
+                    <Route path='contact' component={Contact} />
+                    <Route path='schedule' component={Schedule} >
+                        <IndexRoute component={EventList} />
+                        <Route path=':id' component={(match) => <EventSingle id={match.params.id} />} />
```
Comment (ayc92) [inline web/js/main.jsx]: Seems fine to me haha

I haven't really read the ReactRouter docs though.

File: web/js/components/Schedule/EventSingle.jsx [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+                {date.format("DD")}
+            </div>
+        </div>
+        <div className="event-item__info">
+            Sorry, there are no events scheduled for {date.format("MMMM DD, YYYY")}.
+        </div>
+    </div>
+);
+
+const EventInfo = ({ event }) => (
```
Comment (ayc92) [inline web/js/components/Schedule/EventSingle.jsx]: Sure :)

Either seems fine

File: web/js/components/Schedule/EventMonthItem.jsx (hunk 8 above)
Comment (kamiyo) [inline web/js/components/Schedule/EventMonthItem.jsx]: It was a styling thing but I worked it out.

File: web/js/components/App/App.jsx (hunk 12 above)
Comment (kamiyo) [inline web/js/components/App/App.jsx]: If we leave it the way it was before, the whole container would fade out and in (including the calendar widget on the side). In the case of the individual date navigation, we don't want the whole container to fade, just the sub container. I'm wondering if that means the whole container is rerendering when we don't need it to. You can revert this change to see what I mean.

File: web/js/components/Schedule/EventList.jsx (hunk 13 above)
Comment (kamiyo) [inline web/js/components/Schedule/EventList.jsx]: Yea but before being unmounted, somehow props are being updated.
